pypesto/hierarchical/optimal_scaling_approach/optimal_scaling_problem.py

```diff
@@ -0,0 +1,530 @@
+                    + self.groups[group]['num_datapoints']
+                    + self.groups[group]['num_categories'],
+                ] = -1
+                data_idx += 1
+
+                # x_upper_i - x_lower_{i+1} <= 0
+            if cat_idx == 0:  # - 1:
+                constr[
+                    2 * self.groups[group]['num_datapoints'] + cat_idx,
+                    self.groups[group]['num_datapoints'] + cat_idx,
+                ] = -1
+            else:
+                constr[
+                    2 * self.groups[group]['num_datapoints'] + cat_idx,
+                    self.groups[group]['num_datapoints'] + cat_idx,
+                ] = -1  # + 1] = -1
+                constr[
+                    2 * self.groups[group]['num_datapoints'] + cat_idx,
+                    self.groups[group]['num_datapoints']
+                    + self.groups[group]['num_categories']
+                    + cat_idx
+                    - 1,
+                ] = 1  # + cat_idx] = 1
+
+            constr[
+                2 * self.groups[group]['num_datapoints']
+                + self.groups[group]['num_categories']  # - 1
+                + cat_idx,
+                self.groups[group]['lb_indices'][cat_idx],
+            ] = 1
+            constr[
+                2 * self.groups[group]['num_datapoints']
+                + self.groups[group]['num_categories']  # - 1
+                + cat_idx,
+                self.groups[group]['ub_indices'][cat_idx],
+            ] = -1
+
+        return constr
+
+    def initialize_w(self, gr):
+        """Initialize the weight matrix W for the group 'gr'."""
+        weights = np.diag(
+            np.block(
+                [
+                    np.ones(self.groups[gr]['num_datapoints']),
+                    np.zeros(2 * self.groups[gr]['num_categories']),
+                ]
+            )
+        )
+        return weights
+
+    def get_w(self, gr, y_sim_all):
+        """Return the weight matrix W of the group 'gr'."""
+        weights = np.diag(
+            np.block(
+                [
+                    np.ones(self.groups[gr]['num_datapoints'])
+                    / (np.sum(np.abs(y_sim_all)) + 1e-8),
+                    np.zeros(2 * self.groups[gr]['num_categories']),
+                ]
+            )
+        )
+        return weights
+
+    def get_wdot(self, gr, y_sim_all, sy_all):
+        """Return the derivative of the weight matrix of a group with respect to a outer parameter."""
+        weights = np.diag(
+            np.block(
+                [
+                    np.ones(self.groups[gr]['num_datapoints'])
+                    * (
+                        -1
+                        * np.sum(sy_all)
+                        / ((np.sum(np.abs(y_sim_all)) + 1e-8) ** 2)
+                    ),
+                    np.zeros(2 * self.groups[gr]['num_categories']),
+                ]
+            )
+        )
+        return weights
+
+    def get_w_squared(self, gr, y_sim_all):
```

Comment:
what does the function calculate (informal description)?

Reply:
I've tested some alternative weights of the objective function, and they were not performing as well, so I didn't include them in the implementation of the solver, but I left the weight calculating functions.

I'm going to remove them as they're not being used currently and are easy to implement later if needed.

Reply:
Ah, if these are just alternative weighting functions, also fine to keep them included, and clearly mark which one is used, and which ones are just alternatives.